### ISSUE TITLE
Ambiguity in recurrent neural network training

### ISSUE THREAD

**Issue by amb83** [FluxML/Flux.jl#1528]
(This question may be more a lack of understanding on my part, rather than an issue with Flux, so apologies if this is the wrong place to bring it up.)

I'm using Flux to learn about neural networks.  According to the documentation, each execution of `train!` loops through each `d in data`.  For a conventional NN based on `Dense` -- let's say with a one-dimensional input and output, i.e. with one feature -- this is easy to understand.  Each element in `data` is a pair of single numbers, an independent sample of 1-d input/output values.  `train!` does forward- and backpropagation on each pair of 1-d samples one at a time.  In the process, the `loss` function is evaluated on each sample.  (Do I have this right?)

My question is: how does this extend to a recurrent NN?  Take the case of an RNN with 1-d (i.e. one feature) input and output.  It seems like there's some ambiguity in how to structure the input and output data, and the results change based on the structure.   As one example:

```
x = [[1], [2], [3]]
y = [4, 5, 6]
data = zip(x, y)
m = RNN(1, 1)
opt = Descent()
loss(x, y) = sum((Flux.stack(m.(x), 1) .- y) .^ 2)
train!(loss, params(m), data, opt)
```
(`loss` function taken from: https://github.com/FluxML/Flux.jl/blob/master/docs/src/models/recurrence.md)

In this example, when `train!` loops through each sample (`for d in data`), each value of `d` is a pair of single values from `x` and `y`, e.g. `([1], 4)`.  `loss` is evaluated based on these single values.  This is the same as in the `Dense` case.

On the other hand, consider:

```
x = [[[1], [2], [3]]]
y = [[4, 5, 6]]
m = RNN(1, 1)
opt = Descent()
loss(x, y) = sum((Flux.stack(m.(x), 1) .- y) .^ 2)
train!(loss, params(m), zip(x, y), opt)
```
Note that the only difference here is that `x` and `y` are nested in an extra pair of square brackets.  As a result there's only one `d in data`, and it's a pair of sequences: `([[1], [2], [3]], [4, 5, 6])`.  `loss` can be evaluated on this version of `d`, and it returns a 1-d value, as required for training.  But the value returned by `loss` is different than in any of the three results from the previous case, so the training process turns out differently.

The point is that both structures are valid in the sense that `loss` and `train!` handle them without error.  Conceptually, I can make an argument for both structures being correct.  But the results are different, and I assume that only one way is right.  In other words, for training an RNN, should each `d in data` be a whole sequence, or a single element from a sequence?

**Comment by ToucheSir** [comment]
I'd highly recommend reading through https://fluxml.ai/Flux.jl/dev/models/recurrence/. We hope to get those on the stable site soon.

The easiest way to understand which is "right" is to see what each sample actually looks like (unfortunately `train!` kind of obscures this):
```juliarepl
julia> x = [[1], [2], [3]]; y = [4, 5, 6];

julia> zip(x, y) |> first
([1], 4)

julia> x = [[[1], [2], [3]]]; y = [[4, 5, 6]];

julia> zip(x, y) |> first
([[1], [2], [3]], [4, 5, 6])
```
In the first case, the input is a scalar sequence of length 1 (size `0x1`) and the target is a scalar (size `0`) **. In the second case, the input is a single-feature (note: not scalar) sequence of length 3 (size `(0x1)x3`) and the target is a scalar sequence of length 3 (size `0x3`). So it's not surprising that you see different results, because the two datasets are representing semantically different things!

** here I use `0` to represent the effective dimensionality of a scalar and `x` to mean dimensions rather than multiplication.

**Comment by amb83** [comment]
Hi, thanks for the reply.  I realized my perspective in the original question is off base.  The reason the output of `train!` is different for both input structures is because the `loss` function I used works on both, but with different results.  Specifically, `sum` makes the output of `loss` a scalar for both scalar and sequence inputs.  So it's not a matter of either one being correct.  What this means, I think, is that it's really a design decision.  That is, do you want to train based on the loss being calculated over the whole sequence, or on each point in the sequence one at a time?

But even if that's a fair way to look at things, it still seems like there should be a preferred or recommended way to do it -- training on a whole sequence vs. training on each point.  In the explanation you linked, it uses whole sequences as samples -- the second case I described in my first post (and this approach seems more intuitive).  On the other hand, when I was playing around with an example problem (trying to teach a model sin()), I got better performance with the first approach (training on one point at a time).

(By the way, this is all assuming that I have the right idea now, and that there's not something going on in the background of `train!` that does, in fact, make one approach or the other incorrect.)

**Comment by DhairyaLGandhi** [comment]
`train!` doesn't do anything magical, its your typical

```Julia
for d in data
  gs = gradient(...)
  update!(...)
```

I;m hoping we can close this now?

**Comment by ToucheSir** [comment]
> it uses whole sequences as samples 

To clarify here, this is not possible with Flux. Generally speaking, `Recur` can't take data for more than timestep at a time. You can pass it a single scalar, a batch of samples with one feature or a batch of samples of with multiple features. Note how all of these input formats cover only _one_ timestep. That's also the reason why map/broadcast is used when working with Flux RNNs&mdash;the data for each timestep is distinct and needs to be feed to the network separately.

**Comment by DhairyaLGandhi** [comment]
Fixing that is definitely something on the roadmap, and we used to have some hooks in the long lone ago, but we had reliability concerns at that point. Thanks for the discussion! And please feel free to reopen if deemed like it merits further work.
